qiskit/circuit/quantumcircuit.py

```diff
@@ -1290,67 +1301,68 @@ def qasm(self, formatted=False, filename=None):
+                                                             bit_labels[qubit],
+                                                             bit_labels[clbit])
+            else:
+                # decompose gate using definitions if they are not defined in OpenQASM2
+                if instruction.name not in self.qelib1_gate_names:
+                    if instruction not in self.existing_composite_circuits:
+                        custom_instr_names = [instruction.name for instruction
+                                              in self. existing_composite_circuits]
+                        if instruction.name in custom_instr_names:
+                            # append instruction id to name to make it unique
+                            instruction.name += f'_{id(instruction)}'
+                        self.existing_composite_circuits.append(instruction)
 
                 # Insert qasm representation of the original instruction
-                string_temp += "%s %s;\n" % (instruction.qasm(),
-                                             ",".join([bit_labels[j]
-                                                       for j in qargs + cargs]))
-            else:
-                string_temp += "%s %s;\n" % (instruction.qasm(),
-                                             ",".join([bit_labels[j]
-                                                       for j in qargs + cargs]))
-            if instruction.name == 'unitary':
-                unitary_gates.append(instruction)
+                self.qasm_string_temp += "%s %s;\n" % (instruction.qasm(),
+                                                       ",".join([bit_labels[j]
+                                                                 for j in qargs + cargs]))
 
-        # this resets them, so if another call to qasm() is made the gate def is added again
-        for gate in unitary_gates:
-            gate._qasm_def_written = False
+        # insert gate definitions
+        self._insert_composite_gate_definition_qasm()
 
         if filename:
             with open(filename, 'w+') as file:
-                file.write(string_temp)
+                file.write(self.qasm_string_temp)
             file.close()
 
         if formatted:
             if not HAS_PYGMENTS:
                 raise ImportError("To use the formatted output pygments>2.4 "
                                   "must be installed. To install pygments run "
                                   '"pip install pygments".')
-            code = pygments.highlight(string_temp,
+            code = pygments.highlight(self.qasm_string_temp,
                                       OpenQASMLexer(),
                                       Terminal256Formatter(style=QasmTerminalStyle))
             print(code)
             return None
         else:
-            return string_temp
+            return self.qasm_string_temp
+
+    def _insert_composite_gate_definition_qasm(self):
```

Comment:
Maybe this, as well as `_get_composite_circuit_qasm_from_instruction` can be private module-level functions (same file as `QuantumCircuit` but not inside the class) to keep the code cleaner and easier to read. That would mean that instead of saving the qasm string to an internal variable you would have to pass it around. That's arguably a bit more effort to program but I think it's easier to maintain and read 🙂

Reply:
Sure, let me work on that.